File: SetCover.py
import time

def leer_archivo(nombre_archivo):
    archivo = open(nombre_archivo, 'r')
    lineas = archivo.readlines()
    archivo.close()

    primera_linea = lineas[0].split()
    n = int(primera_linea[0])
    m = int(primera_linea[1])

    subconjuntos = []
    
    for i in range(1, m + 1):
        numeros_texto = lineas[i].split()
        conjunto = set()
        for num in numeros_texto:
            conjunto.add(int(num))
        subconjuntos.append(conjunto)
            
    return n, m, subconjuntos

def greedy_set_cover(universo, subconjuntos):
    no_cubiertos = set(universo)
    seleccion = []
    
    while len(no_cubiertos) > 0:
        mejor_indice = -1
        mejor_cobertura = 0
        mejor_subconjunto = set()
        
        for i in range(len(subconjuntos)):
            elementos_utiles = no_cubiertos.intersection(subconjuntos[i])
            cobertura_actual = len(elementos_utiles)
            
            if cobertura_actual > mejor_cobertura:
                mejor_cobertura = cobertura_actual
                mejor_indice = i
                mejor_subconjunto = subconjuntos[i]
                
        if mejor_indice == -1:
            break
            
        seleccion.append(mejor_indice)
        
        for elemento in mejor_subconjunto:
            if elemento in no_cubiertos:
                no_cubiertos.remove(elemento)
        
    return seleccion

def resolver(indice, seleccion_actual, elementos_cubiertos, universo, subconjuntos, mejor_solucion, mejor_tamano):
    if len(elementos_cubiertos) == len(universo):
        if len(seleccion_actual) < mejor_tamano:
            mejor_tamano = len(seleccion_actual)
            mejor_solucion = []
            for num in seleccion_actual:
                mejor_solucion.append(num)
        return mejor_solucion, mejor_tamano
    
    if indice == len(subconjuntos):
        return mejor_solucion, mejor_tamano
        
    mejor_solucion, mejor_tamano = resolver(indice + 1, seleccion_actual, elementos_cubiertos, universo, subconjuntos, mejor_solucion, mejor_tamano)

    nuevos_cubiertos = set(elementos_cubiertos)
    for elemento in subconjuntos[indice]:
        nuevos_cubiertos.add(elemento)
        
    seleccion_actual.append(indice)
    mejor_solucion, mejor_tamano = resolver(indice + 1, seleccion_actual, nuevos_cubiertos, universo, subconjuntos, mejor_solucion, mejor_tamano)
    
    seleccion_actual.pop()
    
    return mejor_solucion, mejor_tamano

def backtracking_set_cover(universo, subconjuntos):
    mejor_solucion = []
    mejor_tamano = 999999

    mejor_solucion, mejor_tamano = resolver(0, [], set(), universo, subconjuntos, mejor_solucion, mejor_tamano)

    return mejor_solucion

nombre_archivo = 'SetCover/caso1.txt'
n, m, subconjuntos = leer_archivo(nombre_archivo)

universo = set()
for i in range(1, n + 1):
    universo.add(i)
    
print(f"Universo: {n} elementos | Subconjuntos disponibles: {m}\n")
    
inicio_greedy = time.time()
solucion_greedy = greedy_set_cover(universo, subconjuntos)
fin_greedy = time.time()
    
print("Solución Mejorada Greedy")
resultado_greedy = []
for i in solucion_greedy:
    resultado_greedy.append(i + 1)
print(f"Subconjuntos elegidos: {resultado_greedy}")
print(f"Cantidad utilizada: {len(solucion_greedy)}")
print(f"Tiempo: {(fin_greedy - inicio_greedy) * 1000:.4f} ms")
print("------------------------------------------------------")
inicio_back = time.time()
solucion_backtracking = backtracking_set_cover(universo, subconjuntos)
fin_back = time.time()
    
print("Solución Base Backtracking")
resultado_back = []
for i in solucion_backtracking:
    resultado_back.append(i + 1)
print(f"Subconjuntos elegidos: {resultado_back}")
print(f"Cantidad utilizada: {len(solucion_backtracking)}")
print(f"Tiempo: {(fin_back - inicio_back) * 1000:.4f} ms")

File: SetCover.js
const fs = require('fs');

let n = 0;
let m = 0;
let subconjuntos = [];

let mejorTamano = 999999;
let mejorSolucion = [];

function leerArchivo(nombreArchivo) {
    let data = fs.readFileSync(nombreArchivo, 'utf8');
    let lineas = data.split('\n');

    let primeraLinea = lineas[0].trim().split(/\s+/);
    n = parseInt(primeraLinea[0]);
    m = parseInt(primeraLinea[1]);

    for (let i = 1; i <= m; i++) {
        let numerosTexto = lineas[i].trim().split(/\s+/);
        let conjunto = new Set();
        
        for (let j = 0; j < numerosTexto.length; j++) {
            conjunto.add(parseInt(numerosTexto[j]));
        }
        
        subconjuntos.push(conjunto);
    }
}

function greedySetCover(universo) {
    let noCubiertos = new Set();
    for (let elemento of universo) {
        noCubiertos.add(elemento);
    }
    
    let seleccion = [];

    while (noCubiertos.size > 0) {
        let mejorIndice = -1;
        let mejorCobertura = 0;
        let mejorSubconjunto = new Set();

        for (let i = 0; i < subconjuntos.length; i++) {
            let coberturaActual = 0;
            
            for (let elemento of subconjuntos[i]) {
                if (noCubiertos.has(elemento)) {
                    coberturaActual++;
                }
            }

            if (coberturaActual > mejorCobertura) {
                mejorCobertura = coberturaActual;
                mejorIndice = i;
                mejorSubconjunto = subconjuntos[i];
            }
        }

        if (mejorIndice === -1) {
            break;
        }

        seleccion.push(mejorIndice);

        for (let elemento of mejorSubconjunto) {
            noCubiertos.delete(elemento);
        }
    }
    return seleccion;
}

function resolver(indice, seleccionActual, elementosCubiertos) {
    if (elementosCubiertos.size === n) {
        if (seleccionActual.length < mejorTamano) {
            mejorTamano = seleccionActual.length;
            
            mejorSolucion = [];
            for (let i = 0; i < seleccionActual.length; i++) {
                mejorSolucion.push(seleccionActual[i]);
            }
        }
        return;
    }

    if (indice === subconjuntos.length) {
        return;
    }

    resolver(indice + 1, seleccionActual, elementosCubiertos);

    let nuevosCubiertos = new Set();
    for (let elemento of elementosCubiertos) {
        nuevosCubiertos.add(elemento);
    }
    for (let elemento of subconjuntos[indice]) {
        nuevosCubiertos.add(elemento);
    }

    seleccionActual.push(indice);
    resolver(indice + 1, seleccionActual, nuevosCubiertos);
    
    seleccionActual.pop();
}

leerArchivo('caso1.txt');

let universo = new Set();
for (let i = 1; i <= n; i++) {
    universo.add(i);
}

console.log("Universo: " + n + " elementos | Subconjuntos disponibles: " + m + "\n");

let inicioGreedy = performance.now();
let solGreedy = greedySetCover(universo);
let finGreedy = performance.now();

console.log("Solucion Mejorada Greedy");
let resultadoGreedy = [];
for (let i = 0; i < solGreedy.length; i++) {
    resultadoGreedy.push(solGreedy[i] + 1);
}
console.log("Subconjuntos elegidos: [" + resultadoGreedy.join(", ") + "]");
console.log("Cantidad utilizada: " + solGreedy.length);
let tiempoGreedy = finGreedy - inicioGreedy;
console.log("Tiempo: " + tiempoGreedy.toFixed(4) + " ms");

mejorTamano = 999999;
mejorSolucion = [];

let inicioBack = performance.now();
resolver(0, [], new Set());
let finBack = performance.now();
console.log("------------------------------------------------------");
console.log("Solucion Base Backtracking");
let resultadoBack = [];
for (let i = 0; i < mejorSolucion.length; i++) {
    resultadoBack.push(mejorSolucion[i] + 1);
}
console.log("Subconjuntos elegidos: [" + resultadoBack.join(", ") + "]");
console.log("Cantidad utilizada: " + mejorSolucion.length);
let tiempoBack = finBack - inicioBack;
console.log("Tiempo: " + tiempoBack.toFixed(4) + " ms");